wrangler/src/ctrlAccounts/heal.js

```diff
@@ -1,182 +1,140 @@
-    const existingSigners = chain(account.data) 
-    .map((value, key) => {
-      value = Buffer.from(value, 'base64').toString('utf8')
-
-      const signer = find(account.signers, {key: value})
-
-      return key.indexOf('TSS') === 0 && signer ? {
-        turret: key.replace('TSS_', ''),
-        signer: value,
-        weight: signer.weight
-      } : null
-    })
-    .compact()
-    .value()
-
-    return {
-      requiredThreshold: account.thresholds.high_threshold,
-      existingSigners
+  const { TX_FUNCTIONS, TURRET_ADDRESS, STELLAR_NETWORK, HORIZON_URL, META } =
+    env;
+  const feeToken = request.headers.get('authorization')?.split(' ')?.[1];
+  const { publicKey: authPublicKey } = authTxToken(STELLAR_NETWORK, feeToken);
+  const { addTurret, removeTurret } = await request.json();
+  const { txFunctionHash } = params;
+
+  try {
+    const { id: accountPubKey, sequence } = await getAccount(
+      authPublicKey,
+      HORIZON_URL
+    );
+
+    const stellarToml = await META.get('STELLAR_TOML');
+    const turretToml = parse(stellarToml);
+
+    if (removeTurret === TURRET_ADDRESS || addTurret === TURRET_ADDRESS) {
+      return response.json({
+        error: 403,
+        message: `Turrets may not add or remove themselves from a ctrlAccount`,
+      });
     }
-  })
-
-  // return response.json({requiredThreshold, existingSigners})
-
-  const incomingTurretKeys = await Bluebird.map(
-    chain([
-      {turret: addTurret},
-      ...existingSigners
-    ])
-    .orderBy('weight', 'asc')
-    .uniqBy('turret')
-    .value(), 
-    async (signer) => ({
-      ...signer,
-      ...await fetch(`${HORIZON_URL}/accounts/${signer.turret}`)
-      .then((res) => {
-        if (res.ok)
-          return res.json()
-        throw res
-      })
-      .then((account) => {
-        const { url } = request
-        const { hostname } = new URL(url)
-
-        return (
-          hostname === account.home_domain // Workers can't call themselves
-          ? txFunctionsGet({params: {txFunctionHash}}) 
-          : fetch(`https://${account.home_domain}/tx-functions/${txFunctionHash}`)
-        )
-        .then((res) => {
-          if (res.ok)
-            return res.json()
-          throw res
+
+    const { value, metadata } = await TX_FUNCTIONS.getWithMetadata(
+      txFunctionHash,
+      'arrayBuffer'
+    );
+
+    if (!value)
+      return response.json({
+        status: 404,
+        message: `txFunction could not be found this turret`,
+      });
+
+    const { requiredThreshold, accountSigners } = await getAccountSigners(
+      accountPubKey,
+      HORIZON_URL
+    );
+
+    const incomingTurretKeys = await setAccountTurretKeys(
+      turretToml,
+      accountSigners,
+      addTurret,
+      txFunctionHash,
+      HORIZON_URL
+    );
+
+    const removeSigner = find(incomingTurretKeys, { turret: removeTurret });
+
+    if (!removeSigner || !removeSigner.toml)
+      throw 'Signer is not able to be removed';
```

Comment:
Not super sure I follow here.

Whats the requirement to check both `!removeSigner || !removeSigner.toml`?

I assume `!removeSigner` is to validate that the removeSigner is actually a signer on the account

But, it seems like `!removeSigner.toml` just checks that the `toml` exists?